compiler/damlc/daml-stdlib-src/DA/NonEmpty.daml

```diff
@@ -99,6 +106,28 @@ toList (NonEmpty head tail) = head :: tail
 reverse : NonEmpty a -> NonEmpty  a
 reverse l = let hd::tl = L.reverse $ toList l in NonEmpty with hd; tl
 
+-- | Find an element in a non-empty list.
+find : (a -> Bool) -> NonEmpty a -> Optional a
+find p (NonEmpty head tail)
+  | p head = Some head
+  | otherwise = L.find p tail
+
+-- | The 'deleteBy' function behaves like 'delete', but takes a
+-- user-supplied equality predicate.
+deleteBy : (a -> a -> Bool) -> a -> NonEmpty a -> Optional (NonEmpty a)
```

Comment:
Have you considered returning `[]` instead? `Optional (NonEmpty a)` doesn’t provide any extra info or type safety.

Reply:
Good point. I had not considered it. I think given the  trade-off of unwrapping an `Optional` or recreating the `NonEmpty` I prefer the former. Handling the case where you delete the last element is rarer than deleting an element or deleting none. But, yes, that type does look a bit weird.

Reply:
@cocreature Is this ok with you? It would be nice to get this done.

Reply:
sorry missed that, I think I have a preference for the `[a]` case. The type signature here just looks confusing.

Reply:
Can you say why it looks confusing? After this `delete` one will have to handle either a `None` or a `[]`, and there are plenty of [examples](https://docs.daml.com/daml/stdlib/DA-List.html#function-da-list-elemindex-66976) where we wrap in `Optional` to deal with the "missing" case.

Reply:
Generally, I opt for the simplest type that expresses the properties I care about. In this case `[a]` is objectively simpler than `Optional (NonEmpty a)` (both at the type and at the value level) but the more complex type provides zero extra information.

While Haskell doesn’t have `delete` on `NonEmpty` it has [filter](https://hackage.haskell.org/package/base-4.15.0.0/docs/Data-List-NonEmpty.html#v:filter) which is close enough and follows my suggestion here.

Reply:
I don't think that simplicity is the only thing one should consider. I do think that  `Optional (NonEmpty a)` is more explicit in  expressing to the user that some operation may not return a `NonEmpty a`.  But I won't belabor this.